Improve/fix Latex tool usability

Technius opened the pull request:
* [x] Improve error handling (Fixes #1095 somewhat. Possibly prevents #1100).
* [x] ~Set the rendered font color to the currently selected color in the toolbar~ (should be brought up in a separate issue)
* [x] Render asynchronously to avoid the noticeable pauses on keypresses (~100ms on my computer) 
* [x] ~Since pdflatex is relatively slow, delay rendering until 200ms have passed since the last key was entered~ (doesn't bring a noticeable improvement to the user experience)
* [x] ~Make LaTeX tool selectable, with click to bring up the dialog instead of instantly bringing up the dialog and placing the formula in the top left corner~ (too many changes; will bring up in another issue)
* [x] Disable OK button when the entered formula is invalid
* [x] Display a message that explains why the OK button is disabled
* [x] Generate Latex previews in a temporary directory instead of in the config directory
* [x] Include `amssymb` (#1128)

LittleHuba commented:
Rendering every 200ms is no good idea for battery consumption. But we should do it asynchronously after every keypress.

Technius commented:
To clarify, I meant to delay rendering until 200ms have passed after the last key was entered. I've updated the description to reflect that.

LittleHuba commented on src/control/LatexController.cpp:
```diff
@@ -333,6 +448,12 @@ TexImage* LatexController::loadRendered()
 
+	if (!pdfPath.exists())
+	{
+		g_warning("LaTeX preview PDF file does not exist");
```
Also show this to the user.

Technius replied:
Actually, a huge portion of this file could be refactored and simplified, which would make this check unnecessary.

LittleHuba replied:
You may either clean this up in this commit or in a follow-up. Thats for you to decide. I'm happy either way.

Technius replied:
Refactored in one of the commits I pushed today.

Technius commented on src/control/LatexController.cpp:
```diff
@@ -14,15 +14,29 @@
 const char* LATEX_TEMPLATE_2 =
-"\n\\)\n"
-"\\end{document}\n";
+	"\n\\)}\n"
```
You can't insert newlines in an R string. It would have to be `"\n" R"(\))" "\n"` or something uglier (like splitting it across multiple lines).

LittleHuba replied:
Ahh didn't think of that...

Technius commented on src/control/LatexController.cpp:
```diff
@@ -60,15 +74,13 @@ bool LatexController::findTexExecutable()
+	texContents += this->currentTex;
 	texContents += LATEX_TEMPLATE_2;
 
 	string texFile = texTmp + "/tex.tex";
```
Seems like a good issue to fix. One problem: latex generates multiple files for every PDF (there's a `.aux`, a `.div`, and `.log` file too) which could end up generated a lot of unwanted garbage.

Technius replied:
Looks like an easy way is to dump the files in `g_get_temp_dir()` instead of `.xournalpp`.

LittleHuba replied:
Definitely put these files into the temp directory as OSs will clean that directory up regularily. the `.xournalpp` directory is actually only for settings and files that should survive restarts.

Technius replied:
Just pushed this change.

Technius commented:
Looks like `g_uuid_string_random` was implemented in glib 2.52, but Ubuntu 16.04 has an older version of glib.

LittleHuba commented on src/control/LatexController.cpp:
```diff
@@ -110,147 +129,202 @@ void LatexController::findSelectedTexElement()
+	if (this->isUpdating)
+	{
+		return;
+	}
```
It would be better to abort the current image generation and create a new image with the newer data. That could fix the issue with the generating lag. I also see a problem when you type the last character while the image is currently updating.

Technius replied:
On the first point: this doesn't offer any significant advantage to the user experience. This means that if is the user is rapidly typing in the latex, then the user won't be able to see the preview until he or she is done typing. Also, aborting is a bit hard here with the possible race conditions we can have. From the few hours I've been using the new implementation, the lag hasn't been too noticeable.

On the second point: in the callback, after rendering the preview, we check to see if the latex has been updated since the last time an update was triggered. If it has, we trigger another update (see bottom of `onPdfRenderComplete`).

LittleHuba replied:
Ahh okay then you thought about these cases. All good for me!

Technius commented:
Since the last review:
* Rebased onto master
* Use pid in temporary folder name (fixes an outstanding bug)
* Added `amssymb` to list of packages (addresses new feature request)
* Fixed a regression to the cancel button caused by my refactor
* Some code cleanup

This PR should be ready to merge with some more testing (any help welcome!)

veecue commented:
Great work!
Two minor suggestions:
 * Remove the default x^2
 * Add a small indication that the current code produced a build failure (eg a small red bar)

Rmano commented:
I tried this and it's nicely working. I am a quite heavy user of LaTeX; my normal way of entering formulas is to copy and paste from my documents - so I have a couple of suggestions, that can be seen as enhancements proposal for the future:
1 - think about a button to pause the continuous compilation. When in slower machines and with complex formulas, sometimes the lag when the formula is wrong is quite noticeable.
2 - Ideally, the formula text box should be multiline by default, and maybe you could add another text box (under an "expert" switch, maybe) to add text in the latex preamble (so you can add packages or whatever). 
Thanks!

Technius commented:
It seems like everything is working, so I'm going to merge this PR sooner rather than later. I'm going to be refactoring the code in a later PR as the code is still too complicated for what it's doing.

@Rmano A custom preamble feature is definitely going to be useful; I'll open a issue for that soon.

@veecue There's already some text that indicates when the latex is invalid
![image](https://user-images.githubusercontent.com/1066652/57586642-7d306a80-74ad-11e9-897e-34147d0a2406.png)
but it might be a good idea to make the notice more prominent (or by making the formula input have a red background).

veecue commented:
Ok, I actually didn't see the hint.
I would still remove the x^2 default or at least preselect it so that it vanishes when you start typing.

Technius commented:
Preselection is definitely a good idea. I'll look into it.
